src/main/java/io/supertokens/webserver/api/bulkimport/AddBulkImportUsers.java

```diff
@@ -0,0 +1,124 @@
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
+        JsonArray users = InputParser.parseArrayOrThrowError(input, "users", false);
+
+        if (users.size() > MAX_USERS_TO_ADD) {
+            JsonObject errorResponseJson = new JsonObject();
+            errorResponseJson.addProperty("error", "You can only add 1000 users at a time.");
+            super.sendJsonResponse(400, errorResponseJson, resp);
+            return;
+        }
+
+        AppIdentifier appIdentifier = null;
+        try {
+            appIdentifier = getTenantIdentifierFromRequest(req).toAppIdentifier();
+        } catch (ServletException e) {
+            throw new ServletException(e);
+        }
+
+        TenantConfig[] allTenantConfigs = Multitenancy.getAllTenantsForApp(appIdentifier, main);
+        ArrayList<String> validTenantIds = new ArrayList<>();
+        Arrays.stream(allTenantConfigs)
+                .forEach(tenantConfig -> validTenantIds.add(tenantConfig.tenantIdentifier.getTenantId()));
+
+        JsonArray errorsJson = new JsonArray();
+        ArrayList<BulkImportUser> usersToAdd = new ArrayList<>();
+
+        for (int i = 0; i < users.size(); i++) {
+            try {
+                usersToAdd.add(new BulkImportUser(users.get(i).getAsJsonObject(), validTenantIds, null));
+            } catch (InvalidBulkImportDataException e) {
+                JsonObject errorObj = new JsonObject();
+
+                JsonArray errors = e.errors.stream()
+                        .map(JsonPrimitive::new)
+                        .collect(JsonArray::new, JsonArray::add, JsonArray::addAll);
+
+                errorObj.addProperty("index", i);
+                errorObj.add("errors", errors);
+
+                errorsJson.add(errorObj);
+            } catch (Exception e) {
+                JsonObject errorObj = new JsonObject();
+                errorObj.addProperty("index", i);
+                errorObj.addProperty("errors", "An unknown error occurred");
+                errorsJson.add(errorObj);
+            }
+        }
+
+        if (errorsJson.size() > 0) {
+            JsonObject errorResponseJson = new JsonObject();
+            errorResponseJson.addProperty("error",
+                    "Data has missing or invalid fields. Please check the users field for more details.");
+            errorResponseJson.add("users", errorsJson);
+            super.sendJsonResponse(400, errorResponseJson, resp);
+            return;
+        }
+
+        try {
+            AppIdentifierWithStorage appIdentifierWithStorage = getAppIdentifierWithStorage(req);
+            BulkImportStorage storage = appIdentifierWithStorage.getBulkImportStorage();
+            storage.addBulkImportUsers(appIdentifierWithStorage, usersToAdd);
+        } catch (Exception e) {
+            throw new ServletException(e);
```

Comment:
this doe snot belong in this file. Need a separate file for this (see how other recipes are structured)

Reply:
Fixed.